2016-06-28-Rust-Reach-Fractal.md

```diff
@@ -0,0 +1,91 @@
+
+This is where the Steering Committee comes in. It's our job, as leaders of the project, to determine the vision and create a plan for future development.
+
+For us, this is the first major vision-type decision we've had to make since moving to an all-volunteer open governance model. It's challenging to craft cohesive vision in our limited face-to-face meeting time.
+
+To make a decision of this type, we need to lean back on our mission and core principles, on the logistical implications of our decisions, and on the needs of our community.
+
+Critically, we need to pick projects that are well-scoped and highly focused, so that we can progress measurably with the time and person-power we have available. All of these projects are arguably worthwhile. Which is best?
+
+## Fulfilling the vision
+
+Coming back to the core, we want to build the OSHW movement through our platform. This problem fits the original vision for Tessel 2, to help people of different backgrounds create products quickly and easily.
+
+We've come some of the way towards making Tessel 2 the base for a product you can take to market. But we're not all the way there yet. Here are some of the issues we see:
+
+1. It's not easy for most of our users to take the tools we've created so far and go to market. If you don't come from an electrical engineering background, the next step might be insurmountably daunting. You might not know what things you can afford to remove from the design, or how much efficiency or cost they would save you.
+1. Connected device products tend to exist in a networked configuration, usually involving several low-cost, low-power nodes. This makes Tessel 2 not well suited to a lot of product applications.
+1. You can optimize to a point, but at the end of the day you are running JavaScript. This might not be as efficient or as innately reliable as you want your product to be.
+
+## A path forward: Rust, Reach, and Fractal
+
+Drawing on some exploratory work we did over a year ago, we think we can work on these problems such that their solutions build upon each other. Here's the idea:
+
+1. Work on improved **Rust** support. The Rust language is more efficient and explicit than JavaScript and should serve well in production. The code is overall pretty legible, and it's a welcoming community with good documentation. We've already begun tooling for this on the [tessel-rust](//github.com/tessel/tessel-rust) repo.
+1. Build on this for our **Tessel Reach** project: cheap, low-power BLE boards which break out to Tessel module ports. Tessel Reach boards serve as sensor nodes in a star network and are directed by a central Tessel 2 board. You program Tessel 2 in JavaScript; it compiles and sends as-needed communication in Rust to Reach nodes. We have a prototype of this implemented in Rust as the Gossip protocol. We also have some prototype hardware for this. You can see a bit more detail on [this issue](https://github.com/tessel/project/issues/142).
```

Comment:
@HipsterBrown we went with BLE for the prototype hardware we made months ago for a few reasons:
1. The nordic BLE chip series is well documented and cheap
2. The technology had more momentum and name recognition amongst web developers
3. Google was doing some great work with the Physical Web project and were considering joining forces with them at the time
4. We had more in-house expertise of the BLE protocol than any others
5. BLE is lower power than WiFi so it will last on a battery longer (and the ESP8266 wasn't as popular back then)

That being said, we build prototypes more than 15 months ago so times have changed. Feel free to bring up suggestions at the next Steering Committee meeting.

@Frijol I think "star network" might be too advanced for most people. Just "node in a network" is properly a reasonable level of detail.

> You program Tessel 2 in JavaScript; it compiles and sends as-needed communication in Rust to Reach nodes.

@Frijol I'm not sure what you mean here by sending communication in Rust. I think it's more relevant to say, "the program for your Reach actually runs on a Tessel 2; the Tessel emits primitive module port commands over BLE to the Reach."

> We have a prototype of this implemented in Rust as the Gossip protocol.

It might be more useful to say that it's using the same protocol between the Tessel 2's on-board microcontroller and co-processor rather than calling out Gossip.